### ISSUE TITLE
[Perf] Qwen3-Next MoE: gate and shared_expert_gate are two separate ReplicatedLinear over the same input

### ISSUE THREAD

**Issue by pocharlies** [vllm-project/vllm#54763]
## Summary

In `Qwen3NextSparseMoeBlock`, the routed-expert router and the shared-expert gate are two
independent `ReplicatedLinear` layers that consume the **same** `hidden_states`:

https://github.com/vllm-project/vllm/blob/191cecd51e250de77618dda924359214b4d88119/vllm/model_executor/models/qwen3_next.py#L170-L184

```python
self.gate = ReplicatedLinear(
    config.hidden_size,
    config.num_experts,
    bias=False,
    quant_config=None,
    prefix=f"{prefix}.gate",
)

self.shared_expert_gate = ReplicatedLinear(
    config.hidden_size,
    1,
    bias=False,
    quant_config=None,
    prefix=f"{prefix}.shared_expert_gate",
)
```

Same input, same dtype, `bias=False` and `quant_config=None` on both, differing only in output
width (`num_experts` vs `1`). They are a candidate for a single merged projection of width
`num_experts + 1`, split at the output.

**This is a small win and I want to state the size up front: ~1.5% of decode step time on my
setup, and that figure is an upper bound** (see Caveats). I am filing it because the pattern is
easy to miss and the numbers are cheap to record, not because I think it is urgent.

## Why it costs anything at all

At decode the M dimension is tiny (1, or `1 + num_speculative_tokens` with MTP), so these GEMMs
are purely bandwidth-bound: you read the entire weight matrix to produce a handful of output
values. The existing kernels are already at the roofline on this hardware, so there is nothing to
gain from a "better" kernel:

| shape | measured | roofline @ 241 GB/s | |
|---|---|---|---|
| M=1, K=2560, N=6656 | 250 GFLOP/s | 241 | **100%** |
| M=4, K=2560, N=6656 | 940 GFLOP/s | 964 | **98%** |

The gain from merging is not bandwidth. It is removing one kernel invocation whose fixed cost is
not amortised by a single output column — the `N=1` projection reads only ~5 KB of weights.

## Measurements

Environment:

- Model: `RadixArk/Qwen3.8-Flash-Next-NVFP4` — 48 layers, `hidden_size=2560`, 512 experts,
  top-10, `shared_expert_intermediate_size=640`. It subclasses `Qwen3NextSparseMoeBlock`.
- Hardware: 2x NVIDIA DGX Spark (GB10, `sm_121`), TP=2 + expert parallel over 200G RoCE.
- Image `vllm/vllm-openai:qwen38-flash-next`
  (`sha256:fc120ece0a388cc0aa1caad4a9f1cd92113484ab7ec2fd0efadd62585be05bf8`),
  vLLM `0.1.dev20073+g8e685d198`, torch `2.13.0+cu130`.
- MTP-3 speculative decoding, `FULL_DECODE_ONLY` CUDA graphs.

Microbenchmark of the two projections, separate vs merged, BF16, run inside the serving container
on the same GPU:

| M | N | separate | merged | gain |
|---|---|---|---|---|
| 1 | 512 + 1 | 52.3 us | 28.0 us | **+46.4%** |
| 4 | 512 + 1 | 39.6 us | 20.3 us | **+48.7%** |

Scaled to a decode step:

| | |
|---|---|
| saving per MoE block (M=4) | 19.3 us |
| MoE blocks per forward | 48 |
| saving per decode step | 0.93 ms |
| measured decode step | 61.5 ms |
| **share of step** | **~1.5%** |

For context on where the step actually goes, from a `torch.profiler` capture of a real 150-token
generation (4378 ms of kernel time):

| family | % |
|---|---|
| GEMM / MoE experts | 63.2 |
| other | 22.9 |
| NCCL collectives | 10.2 |
| elementwise / copies | 1.7 |
| embedding / gather | 1.3 |
| norm | 0.6 |
| attention | 0.1 |

## Caveats — please read before acting on the number

1. **My microbenchmark ran in eager mode; the server runs `FULL_DECODE_ONLY` CUDA graphs.** Graph
   replay removes most per-launch overhead, which is precisely what this merge saves. The real
   in-situ gain is therefore **smaller than 1.5%**. I could not isolate it because it sits below
   my measurement noise (sigma ~1.2-1.6 tok/s on ~44 tok/s, i.e. ~3%). Treat 1.5% as an upper
   bound, not a result.
2. I did **not** implement or benchmark the merged version end to end. The above is a
   microbenchmark of the two GEMM shapes plus arithmetic.
3. Numbers come from one MoE geometry (512 experts, `hidden_size=2560`). The relative cost of the
   `N=1` GEMM grows as `num_experts` shrinks, so models with fewer experts would benefit more.
4. My measurements were taken on the Qwen3.8-Flash-Next day-0 image, whose `qwen3_next.py`
   is close to but not identical with current `main` (the surrounding fused-shared-expert logic
   has since changed). The two `ReplicatedLinear` declarations themselves are unchanged.

## Implementation note

The awkward part is not the merge but that the two gates have **different consumers**, a few lines
apart:

```python
self.shared_expert = Qwen3NextMLP(..., expert_gate=self.shared_expert_gate, ...)   # L209
self.experts = FusedMoEFactory(..., gate=self.gate, ...)                           # L219
```

Both expect a callable module. A wrapper that slices a merged output but invokes the merged
projection once per consumer would compute it twice and lose the gain. A correct fix computes the
merged projection once in the block's `forward` and passes the two slices down — a small refactor
of the MoE block rather than a declaration change.

`MergedColumnParallelLinear(..., disable_tp=True)` looks like the right primitive; `disable_tp` is
already supported in `linear.py`. As a data point, the Qwen3.8-Flash-Next day-0 image uses exactly
that pattern for its hyperconnection projections, including padding the merged shape to 16 rows
with the comment *"to ensure good alignment and performant implementation chosen by CuBLAS
heuristics"* — that file is not in this repo, but it suggests the padding detail is worth keeping
in mind if anyone implements this.

Given the size of the win and that refactor, I am not opening a PR. This is here so the option is
on record with numbers attached. Happy to run experiments on this hardware if it helps someone
already working in this code.

**Comment by yangsiqt** [comment]
Thanks for the detailed analysis and measurements. I am interested in working on this optimization.

My understanding is that the correct direction is to compute a merged gate projection once in `Qwen3NextSparseMoeBlock.forward`, then split and pass the routed-expert logits and shared-expert gate values to their respective consumers, rather than wrapping two consumers around a shared linear module.

Before I start an implementation, would maintainers consider this refactor acceptable upstream? Given that the reported microbenchmark gain may be largely hidden by CUDA Graph replay, what level of end-to-end validation would you expect for a PR to be worthwhile: correctness coverage plus per-layer microbenchmarks, or a statistically significant serving-level throughput/latency improvement under a representative Qwen3-Next workload?

If this direction is acceptable, I can prepare a minimal implementation and targeted regression tests first.
